Projects/soupstone.md

```diff
@@ -3,19 +3,95 @@
+
+I have quite a few parts like toggle/momentary switches, analog potentiometers, rotary encoders, and LEDs in my parts drawers. The next step will be building some basic proto boards to decide what is fun, move them to PCBs, and make some initial rough units to play with.
 
+[nRF52840-DK]: https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dk
+[stage zero]: https://github.com/jamesmunns/soupstone/pull/2
 
 ## Areas of research
 
+### No Debugger
 
-## Note logs
+In Rust, it is very common to use an external debugging probe via SWD, like a [J-Link] or [DAPLink]. This makes it really easy to reprogram the device, and even get logs via something like [probe-run] and [defmt]. These are made possible using [probe-rs]'s
+
+On one hand, this is super powerful. It makes a lot of the initial board bringup painless, and every Cortex-M board supports it.
+
+On the other hand, not every chip has as good of a debugger story. ESP32 has a couple of different debugging link types. RISC-V processors don't typically have a standard debug interface yet.
+
+On top of that, boards like the XIAO don't have the debugging pins on the header, which means you need pogo pins or solder wires onto the bottom of the board. For people without a soldering iron or the skills to solder to a board or design their own board, this might not be an option at all.
+
+In short: I want to play around with different ways of iterating on firmware, without using a debugger. I hope this opens up more opportunities when you don't have a chip, or a board, or the materials to be able to lean on the debugger to make magic happen.
+
+I plan on using the built-in USB support of the nRF52840, but a similar approach should be possible over UART, or whatever other interfaces are available.
+
+[J-Link]: https://www.segger.com/products/debug-probes/j-link/
+[DAPLink]: https://daplink.io/
+[probe-run]: https://github.com/knurling-rs/probe-run
+[defmt]: https://github.com/knurling-rs/defmt/
+[probe-rs]: https://probe.rs/
+
+### RAM Loader
 
+I got really in my head about having "the perfect bootloader". It can be really hard to make a "bullet proof" bootloader, and flexibily support flashing to both internal flash, and external flash.
+
+To break the "[analysis paralysis]", I decided to make the dumbest persistent loader I could.
+
+This meant:
+
+* It could only write to RAM, not flash
+* A very simple, USB-serial link, using [postcard]
+* Nothing I write with the loader will survive a reboot
+
+Most of the time when you are writing early code, or rapidly iterating, you don't really care about the code running when it isn't attached to your computer. If you have enough RAM to fit both the application, and the RAM the application needs, that means you can skip the slowest part of "flashing" a device: erasing and writing the flash memory.
```

Comment:
This I think I understand, so nice explanation. 
when you are writing early code / rapidly iterating (.. on this external fidget toy) is that correct to assume?